datafusion/physical-plan/src/joins/utils.rs

```diff
@@ -968,6 +991,78 @@ fn estimate_inner_join_cardinality(
     }
 }
 
+/// Estimates semi join cardinality based on statistics.
+///
+/// The estimation result is either zero, in cases inputs statistics are non-overlapping
+/// or equal to number of rows for outer input.
+fn estimate_semi_join_cardinality(
```

Comment:
Long term it would be really nice to pull these types of calculations into some trait (aka an extensibility API)